### DISCUSSION TITLE
Toujours des difficultés avec la manipulation des blocs

### DISCUSSION THREAD

**Discussion by moimp3** [rawsrc/PhpEcho#24] [Q&A] [answered]
Bonjour rawsrc,
Par de nombreux côtés, j'apprécie PhpEcho.
Après lecture et relecture de ta documentation, je ne comprends toujours pas le principe d'imbrication des blocs.
Dans ce code, j'obtiens une erreur:
Fichier 'page/form.4.php':
```php
$page = new PhpEcho('layout/main.php');
$page['title']		= 'Contact';
$page['description']= 'Formulaire de contact, étape 3';
$page['language']	= $language;

$body = new PhpEcho('block/form.4.php');
$body['header']		= new PhpEcho('block/header.php');
$body['language']	= $language;
$body['menu']		= new PhpEcho('block/menu.php');
//$body['errors']		= '';

$type = (int) $_SESSION['post'][1]['type'];
$form		= 20 + $type;
$session	= $_SESSION['post'][$form];

switch($type)
{
	case 1:
		$block['subForm'] = new PhpEcho('block/subForm.41.php');
		break;
	case 2:
		$block['subForm'] = new PhpEcho('block/subForm.42.php');
		break;
}

switch($form)
{
	case 21:
		break;
	case 22:
		switch($session['gender'])
		{
			case 'F':
				$gender	= "Madame";
				break;
			case 'M':
				$gender	= "Monsieur";
				break;
		}
		foreach($session as $key=>$string)
		{
			$block[$key] = $session[$key];
		}
		break;
}

$block['fullname'] = $gender . ' ' . $session['firstname'].' '.$session['familyname'];

$body['content'] = $block;
$page['body']		= $body;
echo $page;
```
Fichier conteneur 'block/form.4.php':
```php
<body>
	<?= $this['header'] ?>
	<h2>Contact&nbsp;: étape 4</h2>
	<div class="mainSection">
		<?= $this['menu'] ?>
		<article id="right">
			<?= $this['content']; ?>
		</article>
	</div>
</body>
```
Bloc enfant 'block/subForm.42.php':
```html
<h3>Détails de votre message</h3>

<p>Vous êtes une organisation (entreprise, association, syndicat, etc.).</p>

<fieldset><legend>Vous</legend>
	<label>Nom&nbsp;:&nbsp;<strong><?= $this['fullname'] ?></strong></label><br>
	<label>Fonction&nbsp;:&nbsp;<strong><?= $this['role'] ?></strong></label><br>
	<label>Tél. direct&nbsp;:&nbsp;<strong><?= $this['directDial'] ?></strong></label><br>
	<label>Courriel&nbsp;:&nbsp;<strong><?= $this['email'] ?></strong></label><br>
</fieldset>
```
Avec ce code, j'obtiens une erreur _Array to string conversion in block\form.4.php on line 7_

**Comment by rawsrc** [comment, marked as answer]
Salut moimp

C'est tout à fait normal l'erreur que tu obtiens, regarde ce que tu fais avec `$block` : 
En 1 : tu codes `$block['subForm'] = new PhpEcho(...);` donc `$block` n'est pas une instance de `PhpEcho`, c'est `$block['subForm']` qui l'est (première erreur), `$block` est un tableau en l'occurrence...
En 2 : tu codes `$block['fullname'] = $gender . ' ' . $session['firstname'].' '.$session['familyname'];`, ce qui exclut d'emblée la possibilité d'avoir un tableau de `PhpEcho`. Un tableau d'instances de `PhpEcho` est géré par le moteur et au rendu il est transformée en chaîne de caractères. Dès que tu mélanges dans un tableau des valeurs de type différents, le rendu du tableau dans la page doit être fait à la main. Tu as une instance de `PhpEcho` à la clé `subForm` et une `string` à la clé `fullname`... Rendu à la mano obligatoire.

Essaie avec ceci : 
```php
$page = new PhpEcho('layout/main.php');
$page['title'] = 'Contact';
$page['description'] = 'Formulaire de contact, étape 3';
$page['language'] = $language;

$body = new PhpEcho('block/form.4.php');
$body['header']	= new PhpEcho('block/header.php');
$body['language'] = $language;
$body['menu'] = new PhpEcho('block/menu.php');

$type = (int) $_SESSION['post'][1]['type'];
$form = 20 + $type;
$session = $_SESSION['post'][$form];

$num_form = match ($type) {
    2 => 42,
    default => 41, // j'ai joué aux devinettes
};

$sub_form = new PhpEcho("block/subForm.{$num_form}.php");

if ($form === 22) {
    $gender = match ($session['gender']) {
        'F' => 'Madame',
        default => 'Monsieur', // parait logique
    };

    foreach ($session as $k => $v) {
        $sub_form[$k] = $v;
    }

    $sub_form['fullname'] = $gender . ' ' . $session['firstname'].' '.$session['familyname'];
} else {
    $sub_form['fullname'] = '';
}

$body['content'] = $sub_form;
$page['body'] = $body;

echo $page;
```

rawsrc

PS : Pense à indiquer ta version de PHP et de PhpEcho 😉

**Reply by moimp3** [reply]
C'est bon. Merci pour ton aide.

**Comment by rawsrc** [comment]
Salut moimp

j'ai oublié de te conseiller pour les blocks répétitifs et/ou inclus d'utiliser dans ton rendu les méthodes prévues pour ces traitements comme `renderByDefault(...)` ou `addBlock(...)` ou `renderBlock(...)`.
Il est plus aisé de lire l'arborescence des blocs directement dans le rendu que de la déduire à partir des fichiers de construction.
Je te conseille également d'utiliser le `ViewBuilder`  qui te permet de ranger ton code dans une instance de classe avec tous les avantages que cela procure.

rawsrc

**Comment by moimp3** [comment]
Salut,
Malgré tous tes efforts et les miens, je suis complètement réfractaire à ta logique d'imbrication des blocs.
Dans l'exemple qui suit et malgré toutes tes explications précédentes, je n'arrive pas à remplacer la partie en commentaire par un nouveau bloc 'block/appSirep'. Je n'arrive pas à créer la variable `$block['text']` pour qu'elle soit accessible avec $this['text'].
C'est vraiment dommage car je vois l'intérêt réel de ton travail.
Je ne comprends pas non plus comment utiliser les fonctions renderBlock() et addBlock().
```php
			$block['style']		= 'background-color:#b0cf77';
			/*
			$block['imgSrc']	= 'images/logoSirep.png';
			$block['goal']		= 'Pour les entreprises';
			$block['title']		= 'La prospection commerciale';
            $block['text'] = <<<'html'
				<ul>				
					<li>Assurez le suivi commercial de vos clients et prospects</li>
					<li>Gérez les rappels de vos clients</li>
					<li>Notez les résumés de vos contacts</li>
					<li>Fonctions spécifiques B to B (d'entreprise à entreprises)</li>
				</ul>               
			html;
			$li[] = "Assurez le suivi commercial de vos clients et prospects";
			$li[] = "Gérez les rappels de vos clients";
			$li[] = "Notez les résumés de vos entretiens";
			$li[] = "Fonctions spécifiques B to B (d'entreprise à entreprises)";
			*/
			$block['text']		= new PhpEcho('block/appSirep.php');

```

Non, je n'ai rien lâché du tout. J'ai contourné le problème (politique de l'autruche) pour avancer dans mon projet.
Oui en tant qu'autodidacte, j'ai des lacunes mais peut-être moins en POO qu'ailleurs.
J'ai des difficultés à comprendre le rôle de tes outils `renderByDefault(...)` ou `addBlock(...)` ou `renderBlock(...)` sans rentrer complètement dans ton code. Aurez-tu des exemples, en plus de ta page d'explications?
En résumé ma difficulté principale concerne l'utilisation de ces fonctions. Mon besoin majeur serait de créer des listes `<li>` de dimension variable en utilisant des array au lieu de string par exemple avec `$body['products'] = myArray` et exploiter le `$this` correspondant avec une boucle foreach dans un bloc.
J'espère que mes explications sont compréhensibles.

**Comment by rawsrc** [comment]
salut moimp,

je me suis permis de ranger un peu tes fils de discussion sinon on ne va pas y arriver

tes vues sont séparées du reste, on est bien d'accord ?

donc pour ton souci de liste, tu dois faire simple : 
- tu passes un tableau de valeurs à la vue
- tu l'affiches dans la vue en utilisant une boucle `foreach`

dans ton fichier de vue, tu dois avoir un truc comme ça :
```php
<ul>
<?php foreach ($this['list'] as $v { ?>
    <li><?= $v ?></li>
<?php } ?>
</ul>
```
et dans la définition de valeurs, tu dois avoir un truc de ce style : 
```php
$li = [
    "Assurez le suivi commercial de vos clients et prospects",
    "Gérez les rappels de vos clients",
    "Notez les résumés de vos entretiens",
    "Fonctions spécifiques B to B (d'entreprise à entreprises)",
];
$block['list'] = $li;
```

**Comment by rawsrc** [comment]
Pour les autres points, je vais faire une mise à jour de `PhpEcho` en version 6.1.0 et dedans j'y ai rajouté la version en français du mode d'emploi 👍

**Comment by moimp3** [comment]
Merci pour tout. Je pense que j'ai bien progressé.
Mon arborescence de répertoires ressemble à la tienne avec les répertoires suivants:
view/fr/block (et autres langues) pour les blocs
view/page pour le code php d'alimentation des blocs (commun à toutes les langues)
J'ai pu réaliser un formulaire multilingue et multipage entièrement avec ton moteur de rendu.
L'exemple que tu m'as donné pour les tableaux m'a bien été. C'est pourtant ce que j'avais essayé de faire mais soit l'âge soit le nez dans le guidon m'ont empêché d'arriver au bout. C'est plus élégant et plus homogène que mon bricolage.
Encore merci parce que je suis satisfait de ton outil.
Il me reste encore à voir les fonctions dont nous avons déjà parlé.

**Comment by moimp3** [comment]
Voici une question concernant ton paragraphe Child blocks / AUTO WIRING VARS:
A la lecture de ton texte, je comprends que ce qui est défini dans le bloc racine est utilisable dans les blocs enfants ce qui s'avère contraire à mon exemple suivant (limité aux seules lignes utiles):
```php
$home = new PhpEcho('layout/main.php');
$home['language']	= $language;
$body = new PhpEcho('block/home.php');
$body['language']	= $language;
$services = [];
$block = new PhpEcho('block/appli.php');
$block['language'] = $language;
$services[]	= $block;
$body['homeFrame']	= new PhpEcho('block/homeFrame.php', ['services' => $services]);
$home['body']		= $body;
echo $home;
```
Je constate que je suis obligé de répéter plusieurs fois la paire clef/valeur de définition de la langue dans chaque bloc et qu'il ne suffit pas de la définir à la racine pour l'utiliser avec $this. Mais ai-je bien compris tes explications?

**Comment by moimp3** [comment]
Erratum sur la fin de mon message précédent:
Erreur: de définition de la langue _dans_ chaque bloc
Correct: de définition de la langue **pour** chaque bloc

**Comment by rawsrc** [comment]
salut moimp

il me semble que tu fais une de ces salades entre `$services` et `block`. J'ai du mal à te suivre.

```php
$body = new PhpEcho('block/home.php', [
    'homeFrame' => new PhpEcho('block/homeFrame.php', ['services' => new PhpEcho('block/appli.php')])
});

echo new PhpEcho('layout/main.php'; [
    'language' => $language,
    'body' => $body
});
```
Pourquoi t'es obligé de redéfinir tes blocs ainsi ?  Normalement, ils devraient être définis de manière statique dans les fichiers de vue (HTML).

**Comment by moimp3** [comment]
Les éléments que je t'ai donné sont un extrait du code que tu m'as donné [ici](https://github.com/rawsrc/PhpEcho/discussions/24#discussioncomment-6769775). A noter toutefois que ce code a été encore étendu. J'ai plusieurs blocs imbriqués (inclus) les uns dans les autres.
Ce qui m'intéresse est de savoir comment éviter de répéter trois fois:
```
$home['language'] = $language;
$body['language'] = $language;
$block['language'] = $language;
```
$home est le parent de $body qui est lui même le parent de $block qui contient lui-même plusieurs blocs dans $services[].
Je pense que tu as les éléments pour me répondre mais dis-moi si tu as besoin du nouveau code complet.

**Comment by rawsrc** [comment]
salut moimp

Tu ne devrais pas gérer le langage ainsi. Pour le peu d'éléments que j'ai, je devine que t'as dû dupliquer le code de tes vues pour chaque langage. C'est une erreur. Tu devrais plutôt utiliser un système de traduction à base de tableaux par exemple (c'est le plus simple) et coder un petit helper dans PhpEcho qui ira chercher la traduction au rendu.

Par exemple, tu définis le langage souhaité comme un paramètre global de PhpEcho et ensuite ton helper va pouvoir identifier dans quel tableau aller piocher les mots clés.

Quand tu codes, tu dois absolument chercher à bannir toute duplication de code, c'est carrément vital. Dès que tu copie/colle avec une légère modification de code, tu dois t'arrêter et réfléchir à la problématique.

rawsrc

**Reply by moimp3** [reply]
Merci pour tous tes conseils. Actuellement, j'ai deux systèmes de traduction.
1-Pour les codes de pages (codes exclusivement en php), j'utilise un fichier json par langue pour tous les messages standard, messages d'erreurs par exemple.
2-Par contre pour tous les blocs de rendu html, j'ai effectivement un fichier par langue. Par exemple, si je prends ton arborescence à la place de template01, j'ai un répertoire par langue. J'étais parti du principe que lorsque j'ai plus de texte que de code, il valait mieux dupliquer les fichiers. En quoi est-ce vital de bannir cette façon de faire?

**Reply by rawsrc** [reply]
Salut,

Pour le point 1 : comment procèdes-tu pour la traduction à la volée ? Une fonction globale ? Tu passes les valeurs déjà traduites à l'instance du bloc ? T'utilises un helper ?

Pour le point 2 : donc à chaque modif de html, tu dois penser à recopier la modif dans chaque fichier vue... C'est casse-gueule. Tu sais l'oubli ou l'erreur est si vite arrivé. Pourquoi ne pas avoir continué à garder ton système de traduction pour cet aspect du code ?

rawsrc

**Reply by moimp3** [reply]
Pour le point 1, j'ai écrit une classe qui prend en paramètre du constructeur le tableau des langues disponibles sur le site. Parmi ces langues, je sélectionne la langue selon plusieurs niveaux de priorité: 1) Le choix de l'utilisateur, 2) Un éventuel enregistrement en cookie, 3) la valeur par défaut du navigateur.

Pour le point 2, je vais tenir compte de ta remarque.

**Comment by moimp3** [comment]
Je n'arrive pas à placer ni à utiliser ma variable '$language' dans un contexte global.
Dans le fichier index, j'ai écrit ceci:
```
$language = 'fr';
PhpEcho::setGlobalParam('language', true);
```
Dans la vue, j'ai ceci: `<?= $this->getAnyParam(name: 'language', seek_order: 'global') ?>`

EDIT: J'ai supprimé la deuxième partie de la question que j'ai résolue.

**Comment by rawsrc** [comment]
salut moimp

il faudrait que tu codes un helper pour tes traductions du style
```php
$i18n = function(string $keyword): string {
    $lang = PhpEcho::getGlobalParam('language');
    // là tu codes la partie en charge d'extraire la traduction en fonction de la langue et du mot clé du dictionnaire
    $translated = '...';

    return $translated;
};
PhpEcho::addHelper('i18n', $i18n);
```
Ensuite dans une vue tu n'as plus qu'à faire `<?= $this->i18n('mot.cle.dico.traduction') ?>`

C'est aussi simple que ça 👍

rawsrc

**Reply by moimp3** [reply]
Pour conclure, je dirai que ton moteur de rendu est génial. Par contre ta documentation manque cruellement de détails et d'exemples.

**Comment by rawsrc** [comment]
bah, je te dirai que sa simplicité objet est sa force et du coup j'ai du mal à me mettre à la place des débutants en PHP et en POO.
Sous ses airs de légèreté, ce moteur embarque un sacré paquets de fonctionnalités et de concepts qui ne sont pas si facilement compréhensibles par des programmeurs débutants voire même intermédiaires. Je l'admets volontiers.

J'ai essayé de faire un effort dans la version 6.1.1 question manuel mais comme généralement je préfère lire le code source à la documentation pour comprendre ce que fait le programme, j'ai tendance à généraliser un peu surtout quand l'outil tient en une seule classe.

C'est pourquoi je réponds bien volontiers aux questions posées via GitHub ou à celles (plus nombreuses) posées directement par mail.

**Comment by moimp3** [comment]
Erreur incompréhensible dans contexte:
Le code suivant renvoie une erreur inexpliquée: "Undefined property: stdClass::$card on line 6"

```
$card	= new PhpEcho('layout/main.php');
$card['title']		= $textes->card_title;
$card['description']	= $textes->card_description;
$card['language']	= $language;
var_dump($card);	// vars retourne les trois valeurs précédentes
$card['h2']		= $textes->card->h2;	// provoque l'erreur ci-dessus
```
Pourtant $card est bien défini !!!

**Comment by rawsrc** [comment]
Salut,
oui $card est bien défini mais pas $textes->*card*->h2 d'où l'erreur

Martin

**Reply by moimp3** [reply]
$textes est une variable externe à PhpEcho. Elle était bien définie mais avec une faute de frappe: l'écriture correcte était `$textes->card_h2`. Désolé pour le dérangement. Mal réveillé!

**Comment by moimp3** [comment]
Bonjour rawsrc,
Je me trouve face à un message d'erreur que je ne m'explique pas:`BadMethodCallException: unknown.view.file.block/editUser.php in C:\wamp64\www\proginet\appOmnes\vendor\rawsrc\phpecho\PhpEcho.php on line 1017` car le fichier /view/block/editUser.php existe bien.
Le code de la page, construit sur le modèle de plusieurs pages, est le suivant:
```php
<?php declare(strict_types=1);
// EDIT USER

use rawsrc\PhpEcho\PhpEcho;

// check current user as security
if( empty($_SESSION['user']['webmaster']) )
{
	header('Location: index.php');
	exit;
}

$Users = getUsers();
$lstUser = new OptListSelect('user');
$lstUser->addOption( new OptListOption() );
$lstUser->addOption( new OptListOption('0', "&lt;nouveau&gt;") );

if($Users)
{
	foreach($Users as $user)
	{
		$lstUser->addOption( new OptListOption( (string) $user['id'], "{$user['login']}. ({$user['fullname']})") );
	}
}

$user				= new PhpEcho('layout/main.php');
$user['title']		= $textes->head_title;
$user['description']= $textes->head_description;
$user['language']	= 'fr';

$body				= new PhpEcho('block/editUser.php');
$body['errors']		= errorHandler();
$body['language']	= 'fr';

$body['lstUser']	= $lstUser;
// for next three lines, pages are required in index.php and parameters are defined in relevant pages 
$body['header']		= $header;
$body['footer']		= $footer;
$body['menu'] 		= $menu;

$user['body']		= $body;

echo $user;
```

**Comment by rawsrc** [comment]
salut moimp,

je ne vois pas dans ton code
```php
PhpEcho::setTemplateDirRoot('répertoire_racine_pour_toutes_les_vues');
```
c'est normal ?
Sans cette définition, impossible de trouver ton bloc, d'où l'erreur

**Comment by moimp3** [comment]
Bonjour rawsrc,
Le code que je t'ai donné est celui d'un fichier inclus. Le fichier parent débute comme ceci:
```php
session_start();

require_once 'config.php';
require_once 'define.php';
require_once PATH_MODEL;

spl_autoload_register(function (string $className)
{
	foreach ( glob(DIR_CLASS.DIRECTORY_SEPARATOR.'*', GLOB_ONLYDIR) as $dir )
	{
		if( file_exists($dir.DIRECTORY_SEPARATOR.$className.".php") )
			require_once $dir.DIRECTORY_SEPARATOR.$className.".php";
	}
});
require_once DIR_ROOT.'vendor'.DIRECTORY_SEPARATOR.'autoload.php';
require_once PATH_FUNCTIONS;

use rawsrc\PhpEcho\PhpEcho;

// ... Recherche de la langue et du fichier de textes

PhpEcho::setTemplateDirRoot(DIR_ROOT.'view');

```
Voilà pourquoi je ne comprends pas.

**Comment by moimp3** [comment]
Bonjour rawsrc,
Le code que j'ai donné est celui d'un fichier inclus dans `index.php` (index et routeur).
Le `setTemplateDirRoot()` est dans ce fichier parent, d'ailleurs le fichier `layout/main.php` est bien trouvé.

**Comment by rawsrc** [comment]
salut,

si ça fonctionne pour `layout/main.php` ça doit fonctionner pour tout le reste.
le problème se situe ailleurs : nommage de fichier, caractère masqué, problème de casse...
voire même réinitialisation du répertoire racine des vues, ce n'est pas des raisons qui manquent

tu dois investiguer ton code.

**Reply by moimp3** [reply]
Effectivement, j'ai trouvé: Dans l'explorateur Windows, j'avais 'block/editUser..php' au lieu de 'block/editUser.php' (pas très facile à voir quand on a des problèmes de vue)

**Comment by rawsrc** [comment]
euh ton autloader est vachement bancal
tiens du temps de dvp, j'avais écrit un billet sur cette problématique [https://www.developpez.net/forums/blogs/32058-rawsrc/b5109/autoloader/](https://www.developpez.net/forums/blogs/32058-rawsrc/b5109/autoloader/)
